File: crates/yokan/demo/keys.py
# /// script
# requires-python = ">=3.14"
# ///
"""Keys and the clipboard. `shortcut(chord, handler)` declares a chord
the app answers, and `on_key(handler)` sees every key as the chord it
was. The chord is spelled the way the platform spells it (`cmd+s`,
`shift-tab`, `ctrl+alt+k`); a headless script presses one with
`key:cmd+s`, so a shortcut is checked by the gate like any other
interaction. `clipboard.set_text` / `get_text` copy and paste: a
window exchanges the text with every other application, a headless
run keeps it to itself. `menu_item(menu, name, handler)` puts the
same handlers in the application's menu bar, and a script picks one
with `menu:Save`.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import (  # noqa: E402
    button,
    column,
    menu_item,
    on_key,
    row,
    run,
    shortcut,
    State,
    text,
)
from yokan import clipboard  # noqa: E402

count: State[int] = State(0)
saved: State[int] = State(0)
last: State[str] = State("-")
pasted: State[str] = State("(nothing)")


def bump():
    count.set(count() + 1)


def save():
    saved.set(count())


def clear():
    count.set(0)
    saved.set(0)


def typed(key: str):
    last.set(key)


def copy_count():
    clipboard.set_text(f"count={count()}")


def paste():
    pasted.set(clipboard.get_text())


menu_item("Count", "Save", save)
menu_item("Count", "Clear", clear)

shortcut("cmd+s", save)
shortcut("cmd+shift+r", clear)
shortcut("cmd+shift+c", copy_count)
shortcut("cmd+shift+v", paste)
on_key(typed)


def view():
    with column(spacing=8, padding=12):
        text(f"count: {count()}  saved: {saved()}")
        text(f"last key: {last()}")
        text(f"pasted: {pasted()}")
        with row(spacing=6):
            button("+1", on_click=bump)
            button("save", on_click=save)
            button("copy", on_click=copy_count)
            button("paste", on_click=paste)


if __name__ == "__main__":
    run(view, title="keys")

File: rakugan/demo/keys.pl
# The keyboard as a set of chords, and the same handlers in the
# application's menu bar. A script presses one with `key:cmd+s` and
# picks one with `menu:Save`.
use Rakugan;

class Keys {
    use Rakugan;
    field $count  = 0;
    field $saved  = 0;
    field $last   = "-";
    field $pasted = "(nothing)";

    method save {
        $saved = $count;
    }

    method clear {
        $count = 0;
        $saved = 0;
    }

    method copy_count {
        clipboard_set_text("count=$count");
    }

    method paste {
        $pasted = clipboard_get_text();
    }

    method typed :Sig(Str) ($key) {
        $last = $key;
    }

    method view {
        return column(
            text("count: $count  saved: $saved"),
            text("last key: $last"),
            text("pasted: $pasted"),
            row(
                button("+1",    on_click => sub { $count += 1 }),
                button("save",  on_click => sub { $self->save }),
                button("copy",  on_click => sub { $self->copy_count }),
                button("paste", on_click => sub { $self->paste }),
                spacing => 6,
            ),
            spacing => 8,
            padding => 12,
        );
    }
}

my $app = Keys->new;

menu_item("Count", "Save",  sub { $app->save });
menu_item("Count", "Clear", sub { $app->clear });

shortcut("cmd+s",        sub { $app->save });
shortcut("cmd+shift+r",  sub { $app->clear });
shortcut("cmd+shift+c",  sub { $app->copy_count });
shortcut("cmd+shift+v",  sub { $app->paste });
on_key(sub ($chord) { $app->typed($chord) });

run($app, title => "keys");
